taskhawk/backends/base.py

```diff
@@ -111,9 +116,9 @@ def fetch_and_process_messages(self, num_messages: int = 1, visibility_timeout:
                 continue
             except RetryException as exc:
                 # Retry without logging exception
-                if exc.delay_seconds > 0:
+                if exc.delay_seconds > 0 and exc.metadata is not None:
                     logger.info(f'Retrying with delay {exc.delay_seconds} seconds')
-                    self.extend_visibility_timeout(exc.delay_seconds, queue_message.metadata)
```

Comment:
`queue_message` is a provider specific message type, it doesn't provide metadata directly. I decided to inject metadata to retry exception to avoid major code changes and keep current public methods signature unchanged.

Alternative option would be to build `message: Message` instance using `_build_message()` method or just add optional `visibility_s`: int to param to `nack_message()` method.

Reply:
ugh, I see. Adding `visibility_s` to `nack_message` seems odd, since the fact that you nack by setting visibility to 0 _seems_ like an implementation detail of specific providers (e.g. RabbitMQ has a separate `nack` method (not that we use RabbitMQ)), but given that both AWS and GCP work that way, perhaps just adding `visibility_s` into `nack_message` is better. Alternatively, maybe changing signature of `extend_visibility_timeout` to include `queue_message` object?

That is:
```
diff --git a/docs/usage.rst b/docs/usage.rst
index 1695d56..e96e0d2 100644
--- a/docs/usage.rst
+++ b/docs/usage.rst
@@ -25,7 +25,7 @@ example, request id.
 
 **id**: task identifier. This represents a run of a task.
 
-**platform_metadata**: Platform specific metadata, for example SQS receipt or PubSub ack id. This may be used to
+**provider_metadata**: Platform specific metadata, for example SQS receipt or PubSub ack id. This may be used to
 extend message visibility if the task is running longer than expected using ``taskhawk.extend_visibility_timeout``.
 
 **priority**: the priority a task was dispatched with. This will be same as task's priority, unless priority was
diff --git a/taskhawk/backends/aws.py b/taskhawk/backends/aws.py
index be5834f..b82dc6e 100644
--- a/taskhawk/backends/aws.py
+++ b/taskhawk/backends/aws.py
@@ -177,11 +177,11 @@ class AWSSQSConsumerBackend(TaskhawkConsumerBaseBackend):
         # should operate like a nack https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html#terminating-message-visibility-timeout
         queue_message.change_visibility(VisibilityTimeout=0)
 
-    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata: AWSMetadata) -> None:
+    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata: Optional[AWSMetadata] = None, queue_message: Optional = None) -> None:
         """
         Extends visibility timeout of a message on a given priority queue for long running tasks.
         """
-        receipt = metadata.receipt
+        receipt = metadata.receipt if metadata else queue_message.receipt_handle
         queue_url = self.sqs_client.get_queue_url(QueueName=self.queue_name)['QueueUrl']
         self.sqs_client.change_message_visibility(
             QueueUrl=queue_url, ReceiptHandle=receipt, VisibilityTimeout=visibility_timeout_s
@@ -238,7 +238,7 @@ class AWSSQSConsumerBackend(TaskhawkConsumerBaseBackend):
 
 
 class AWSSNSConsumerBackend(TaskhawkConsumerBaseBackend):
-    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata) -> None:
+    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata: Optional[AWSMetadata] = None, queue_message: Optional = None) -> None:
         raise RuntimeError("invalid operation for backend")
 
     def requeue_dead_letter(self, num_messages: int = 10, visibility_timeout: Optional[int] = None) -> None:
diff --git a/taskhawk/backends/base.py b/taskhawk/backends/base.py
index 1d29a07..fea8704 100644
--- a/taskhawk/backends/base.py
+++ b/taskhawk/backends/base.py
@@ -113,7 +113,7 @@ class TaskhawkConsumerBaseBackend(TaskhawkBaseBackend):
                 # Retry without logging exception
                 if exc.delay_seconds > 0:
                     logger.info(f'Retrying with delay {exc.delay_seconds} seconds')
-                    self.extend_visibility_timeout(exc.delay_seconds, queue_message.metadata)
+                    self.extend_visibility_timeout(exc.delay_seconds, queue_message)
                     # the `continue` below will prevent the `self.delete_message` call from deleting the message from the queue.
                 else:
                     logger.info('Retrying due to exception')
@@ -135,7 +135,7 @@ class TaskhawkConsumerBaseBackend(TaskhawkBaseBackend):
             except Exception:
                 logger.exception('Exception while deleting message', extra={'queue_message': queue_message})
 
-    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata) -> None:
+    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata: Optional = None, queue_message: Optional = None) -> None:
         """
         Extends visibility timeout of a message on a given priority queue for long running tasks.
         """
diff --git a/taskhawk/backends/gcp.py b/taskhawk/backends/gcp.py
index 4e738ad..55e981f 100644
--- a/taskhawk/backends/gcp.py
+++ b/taskhawk/backends/gcp.py
@@ -243,16 +243,17 @@ class GooglePubSubConsumerBackend(TaskhawkConsumerBaseBackend):
     def post_process_hook_kwargs(queue_message: ReceivedMessage) -> dict:
         return {'google_pubsub_message': queue_message}
 
-    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata: GoogleMetadata) -> None:
+    def extend_visibility_timeout(self, visibility_timeout_s: int, metadata: Optional[GoogleMetadata] = None, queue_message: Optional[ReceivedMessage] = None) -> None:
         """
         Extends visibility timeout of a message on a given priority queue for long running tasks.
         """
+        ack_id = metadata.ack_id if metadata else queue_message.ack_id
         if settings.TASKHAWK_SYNC:
             return
         if visibility_timeout_s < 0 or visibility_timeout_s > 600:
             raise ValueError("Invalid visibility_timeout_s")
         self.subscriber.modify_ack_deadline(
-            subscription=self._subscription_path, ack_ids=[metadata.ack_id], ack_deadline_seconds=visibility_timeout_s
+            subscription=self._subscription_path, ack_ids=[ack_id], ack_deadline_seconds=visibility_timeout_s
         )
         self._call_heartbeat_hook()

```

Reply:
changes applied

Reply:
alternative `nack_message()` approach: https://github.com/cloudchacho/taskhawk-python/pull/14